datalad/core/distributed/push.py

```diff
@@ -0,0 +1,726 @@
+            target = set(p.get('remote', None) for p in wannabe_gitpush)
+            # handle case where a pushinfo record did not have a 'remote'
+            # property -- should not happen, but be robust
+            target.discard(None)
+        except Exception as e:
+            lgr.debug(
+                'Dry-run push to determine default push target failed, '
+                'assume no configuration: %s', e)
+            target = set()
+        if not len(target):
+            yield dict(
+                res_kwargs,
+                status='error',
+                message='No push target given, and none could be '
+                        'auto-detected',
+            )
+            return
+        elif len(target) > 1:
+            # dunno if this can ever happen, but if it does, report
+            # nicely
+            yield dict(
+                res_kwargs,
+                status='error',
+                message=(
+                    'No push target given, '
+                    'multiple candidates auto-detected: %s',
+                    list(target),
+                )
+            )
+            return
+        else:
+            # can only be a single one at this point
+            target = target.pop()
+
+    if target not in repo.get_remotes():
+        yield dict(
+            res_kwargs,
+            status='error',
+            message=(
+                "Dataset %s does not know of a sibling '%s' to push to.",
+                ds, target))
+        return
+
+    lgr.debug("Attempt to push to '%s'", target)
+
+    # define config var name for potential publication dependencies
+    depvar = 'remote.{}.datalad-publish-depends'.format(target)
+    # list of remotes that are publication dependencies for the
+    # target remote
+    publish_depends = ensure_list(ds.config.get(depvar, []))
+    if publish_depends:
+        lgr.debug("Discovered publication dependencies for '%s': %s'",
+                  target, publish_depends)
+
+    # anything that follows will not change the repo type, cache.
+    is_annex_repo = isinstance(ds.repo, AnnexRepo)
+
+    # TODO prevent this when `target` is a special remote
+    # (possibly redo) a push attempt to figure out what needs pushing
+    # do this on the main target only, and apply the result to all
+    # dependencies
+    try:
+        wannabe_gitpush = repo.push(
+            remote=target,
+            git_options=['--dry-run'])
+    except Exception as e:
+        lgr.debug(
+            'Dry-run push to check push configuration failed, '
+            'assume no configuration: %s', e)
+        wannabe_gitpush = []
+    refspecs2push = [
+        # if an upstream branch is set, go with it
+        p['from_ref']
+        if ds.config.get(
+            'branch.{}.remote'.format(p['from_ref'][len('refs/heads/'):]),
+            None) == target and ds.config.get(
+            'branch.{}.merge'.format(p['from_ref'][len('refs/heads/'):]),
+            None)
```

Comment:
I'm having a hard time digesting this, so perhaps I'm misreading it, but I don't see why you'd need to index `branch.*.merge` by `len('refs/heads/'):` because you're not comparing it to anything.  And I think indexing `branch.*.remote` by `len('refs/heads/'):` is always wrong; it's a remote or ".".

Reply:
I am not indexing the variable spec, but `p['from_ref']`. I have adjusted that part and added a comment

Reply:
I managed to get very confused.  Sorry for the noise.